File: leetcode-04/solutions/python3/0713.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0

        ans = 0
        prod = 1
        j = 0

        for i in range(len(nums)):
            prod *= nums[i]
            while prod >= k:
                prod /= nums[j]
                j += 1
            ans += i - j + 1

        return ans

File: leetcode-04/solutions/cpp/0713.cpp
class Solution {
 public:
  int numSubarrayProductLessThanK(vector<int>& nums, int k) {
    if (k <= 1) return 0;

    int ans = 0;
    int prod = 1;
    int j = 0;

    for (int i = 0; i < nums.size(); i++) {
      prod *= nums[i];
      while (prod >= k) prod /= nums[j++];
      ans += i - j + 1;
    }

    return ans;
  }
};